onnxruntime/core/providers/qnn/qnn_allocator.cc

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/qnn/qnn_allocator.h"
+
+#include <cassert>
+#include <cstddef>
+#include <algorithm>
+#include <limits>
+
+#include "core/common/common.h"
+#include "core/mlas/inc/mlas.h"  // for MlasGetPreferredBufferAlignment()
+
+namespace onnxruntime::qnn {
+
+namespace {
+
+struct AllocationHeader {
+  static constexpr std::array<char, 8> kAllocationHeaderMarker{'o', 'r', 't', 'a', 'l', 'l', 'o', 'c'};
+
+  // Marker bytes to verify as a sanity check.
+  std::array<char, 8> marker;
+
+  // Pointer to the allocating allocator instance.
+  // Note: A critical assumption here is that the allocating allocator is not destroyed before the allocation is freed.
+  HtpSharedMemoryAllocator* allocator_ptr;
+
+  AllocationHeader(HtpSharedMemoryAllocator* allocator_ptr)
+      : marker{kAllocationHeaderMarker},
+        allocator_ptr{allocator_ptr} {
+  }
+
+  ~AllocationHeader() {
+    marker.fill('\0');
+    allocator_ptr = nullptr;
```

Comment:
Should we limit doing the `fill` to a debug build? not sure how many allocations QNN makes and whether there's any meaningful perf cost.

Reply:
I'm a little hesitant to remove it as it did catch some issues during my testing. maybe we can do that later if it is measured to have a significant performance cost? it is only overwriting the 8 marker bytes.